src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java

```diff
@@ -93,4 +99,139 @@ public static String toQueryString(Map<String, String> params) {
+        if (
+            !stopsMatch(previousItineraryLeg.from, newItineraryLeg.from) ||
+                !stopsMatch(previousItineraryLeg.to, newItineraryLeg.to)
+        ) {
+            return false;
+        }
+
+        // make sure the transit service is the same as perceived by the customer
+        if (
+            !equalsOrPreviousWasNull(previousItineraryLeg.mode, newItineraryLeg.mode) ||
+                !equalsIgnoreCaseOrPreviousWasEmpty(previousItineraryLeg.agencyName, newItineraryLeg.agencyName) ||
+                !equalsIgnoreCaseOrPreviousWasEmpty(
+                    previousItineraryLeg.routeLongName,
+                    newItineraryLeg.routeLongName
+                ) ||
+                !equalsIgnoreCaseOrPreviousWasEmpty(
+                    previousItineraryLeg.routeShortName,
+                    newItineraryLeg.routeShortName
+                ) ||
+                !equalsIgnoreCaseOrPreviousWasEmpty(
+                    previousItineraryLeg.headsign,
+                    newItineraryLeg.headsign
+                ) ||
+                !equalsOrPreviousWasNull(
+                    previousItineraryLeg.interlineWithPreviousLeg,
+                    newItineraryLeg.interlineWithPreviousLeg
+                )
+        ) {
+            return false;
+        }
+
+        // make sure the transit trips are scheduled for the same time of the day
+        if (
+            !timeOfDayMatches(
+                previousItineraryLeg.getScheduledStartTime(),
+                newItineraryLeg.getScheduledStartTime()
+            ) || !timeOfDayMatches(
+                previousItineraryLeg.getScheduledEndTime(),
+                newItineraryLeg.getScheduledEndTime()
+            )
+        ) {
+            return false;
+        }
+
+        // if this point is reached, the legs are assumed to match
+        return true;
+    }
+
+    /**
+     * Checks whether two stops (OTP Places) match for the purposes of matching itineraries
+     */
+    private static boolean stopsMatch(Place stopA, Place stopB) {
+        // stop names must match
+        if (!stopA.name.equalsIgnoreCase(stopB.name)) return false;
+
+        // stop code must match
+        if (!equalsIgnoreCaseOrPreviousWasEmpty(stopA.stopCode, stopB.stopCode)) return false;
+
+        // stop positions must be no further than 5 meters apart
+        if (
+            DistanceUtils.radians2Dist(
+                DistanceUtils.distHaversineRAD(stopA.lat, stopA.lon, stopB.lat, stopB.lon),
+                DistanceUtils.EARTH_MEAN_RADIUS_KM
+            ) * 1000 > 5
+        ) {
+            return false;
+        }
+
+        // if this point is reached, the stops are assumed to match
+        return true;
+    }
+
+    /**
+     * Returns true if previous is null. Otherwise, returns Objects.equals.
+     */
+    private static boolean equalsOrPreviousWasNull (Object previous, Object newer) {
+        return previous == null || Objects.equals(previous, newer);
```

Comment:
I don't really understand the use of this method throughout these checks. Could you explain why it's ok if just previous is null? What if previous is null and newer is not null? Won't that return true and violate our expectation that the values are the same?

Reply:
The idea is that perhaps an agency doesn't include some relevant information in their GTFS such as route long name that might have been there all along as perceived by the user. Once the transit agency does include that info, it could result in the leg not matching when it really shouldn't. So it gives some leniency if a transit agency had a less detailed feed that now has more details.